feature_discovery.md

```diff
@@ -0,0 +1,158 @@
+                "roles": ["player"]
+            }
+        ]
+    }
+}
+```
+
+The `protocols` field is a JSON array of __protocol support descriptor__
+objects that match the query. Each descriptor has a `pid` that contains
+a protocol version (fully qualified message family identifier such as
+`https://didcomm.org/tictactoe/1.0`), plus a `roles`
+array that enumerates the roles the responding agent
+can play in the associated protocol.
+
+Response messages say, "Here are some protocols I support that matched
+your query, and some things I can do with each one."
+
+##### Sparse Responses
+
+Responses do not have to contain exhaustive detail. For example, the following
+response is probably just as good:
+
+```json
+{
+  "type": "https://didcomm.org/discover-features/1.0/disclose",
+  "thread_id": "yWd8wfYzhmuXX3hmLNaV5bVbAjbWaU",
+  "protocols": [
+    {"ptid": "https://didcomm.org/tictactoe/1.0"}
+  ]
+}
+```
+
+The reason why less detail probably suffices is that agents do not need to
+know everything about one another's implementations in order to start an
+interaction--usually the flow will organically reveal what's needed. For
+example, the `outcome` message in the `tictactoe` protocol isn't needed
+until the end, and is optional anyway. Alice can start a tictactoe game
+with Bob and will eventually see whether he has the right idea about
+`outcome` messages.
+
+The missing `roles` in this response does not say, "I support no roles
+in this protocol." It says, "I support the protocol but
+I'm providing no detail about specific roles."
+
+Even an empty `protocols` map does not say, "I support no protocols
+that match your query." It says, "I'm not telling you that I support any
+protocols that match your query." An agent might not tell another that
+it supports a protocol for various reasons, including: the trust that
+it imputes to the other party based on cumulative interactions so far,
+whether it's in the middle of upgrading a plugin, whether it's currently
+under high load, and so forth. And responses to a `discover-features` request are
+not guaranteed to be true forever; agents can be upgraded or downgraded,
+although they probably won't churn in their protocol support from moment
+to moment.
+
+### Privacy Considerations
+
+Because the regex in a `request` message can be very inclusive, the `discover-features`
+protocol could be used to mine information suitable for agent fingerprinting,
+in much the same way that browser fingerprinting works. This is antithetical
+to the ethos of our ecosystem, and represents bad behavior. Agents should
+use `discover-features` to answer legitimate questions, and not to build detailed
+profiles of one another. However, fingerprinting may be attempted
+anyway.
+
+For agents that want to maintain privacy, several best practices are
+recommended:
+
+##### Follow selective disclosure.
+
+Only reveal supported features based on trust in the relationship.
```

Comment:
How do we codify and test for "trust in the relationship"? Having more details here could be useful as to the expected way that an agent would determine whether it should send a subset of features on it's list or the whole list.

Reply:
Why certain feature should be revealed is quite specific to the implementation and the features themselves. I do understand the desire to end up with testable statements. Should discussion of selective disclosure belong in the guide instead of the spec?

Reply:
Agreed on call this week that this would be good to move to the implementation guide. I'm fine with leaving this statement in the spec if you want as well.